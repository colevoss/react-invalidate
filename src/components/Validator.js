// @flow

import { PropTypes, Component } from 'react';

import type {
  ValidatorFunction,
  ValidatorChildFunction,
  ChildParams,
  FieldValidtor,
  ValidationClearer,
} from '../shared/types';

import { ValidatorFunctionPropTypes } from '../shared/propTypes';

type State = {
  isValid: boolean,
  message: ?string,
  lastValidatedValue: ?any,
};

type Props = {
  children: ValidatorChildFunction,
  validators: ValidatorFunction | Array<ValidatorFunction>,
  id: string,
  initialValue?: any,
  validateOnMount: boolean,
};

export default class Validator extends Component {
  constructor(props: Props) {
    super(props);

    this.id = props.id;

    this.state = {
      isValid: true,
      message: null,
      lastValidatedValue: this.props.initialValue,
    };
  }

  _mounted: boolean;
  id: string;
  state: State;
  props: Props;

  componentDidMount() {
    this._mounted = true;

    this.validateOnMount();
  }

  componentWillUnmount() {
    this._mounted = false;
  }

  validateOnMount(): void {
    this.props.validateOnMount && this.validate();
  }

  defaultState(): State {
    return {
      isValid: true,
      message: null,
      lastValidatedValue: this.state.lastValidatedValue,
    };
  }

  clearValidation: ValidationClearer = () => {
    this.setState(this.defaultState());
  };

  validate: FieldValidtor = value => {
    const validators = Array.isArray(this.props.validators)
      ? this.props.validators
      : [this.props.validators];

    const toValidate = value || this.state.lastValidatedValue;

    return Promise.all(validators.map(validator => validator(toValidate)))
      .then(
        // All validators passed
        () => {
          this._mounted &&
            this.setState({
              isValid: true,
              message: null,
              lastValidatedValue: value,
            });

          return true;
        }
      )
      .catch(
        // Message of the first validation to fail
        message => {
          this._mounted &&
            this.setState({
              isValid: false,
              message,
              lastValidatedValue: value,
            });

          return message;
        }
      );
  };

  childParams(): ChildParams {
    return {
      validate: this.validate,
      isValid: this.state.isValid,
      message: this.state.message,
      clearValidation: this.clearValidation,
    };
  }

  render() {
    return this.props.children(this.childParams());
  }
}

Validator.propTypes = {
  children: PropTypes.func,
  validators: ValidatorFunctionPropTypes,
  id: PropTypes.string,
  initialValue: PropTypes.any,
  validateOnMount: PropTypes.bool,
};
