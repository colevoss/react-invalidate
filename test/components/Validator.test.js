import React from 'react';
import { shallow, mount } from 'enzyme';
import Validator from '../../src/components/Validator';

const trueValidator = () => true;
const failedValidator = () => Promise.reject('message');

describe('<Validator />', () => {
  test("renders it child function's component directly", () => {
    const wrapper = shallow(
      <Validator>
        {() => <div className="test-child" />}
      </Validator>
    );

    expect(wrapper.is('.test-child'));
    expect(wrapper).toMatchSnapshot();
  });

  test('passes the correct params to child function', () => {
    shallow(
      <Validator>
        {params => {
          expect(params.isValid).toBeTruthy();
          expect(params.message).toBeNull();
          expect(params.validate).toBeInstanceOf(Function);
          expect(params.clearValidation).toBeInstanceOf(Function);
          expect(params).toMatchSnapshot();

          return <div className="test-child" />;
        }}
      </Validator>
    );
  });

  describe('#validate', () => {
    test('returns Promise<true> if validation passes', () => {
      const wrapper = shallow(
        <Validator validators={trueValidator}>
          {({ validate }) => (
            <input
              type="text"
              onBlur={() => {
                validate().then(val => expect(val).toBeTruthy());
              }}
            />
          )}
        </Validator>
      );

      wrapper.find('input').simulate('blur');
    });

    test('cathes rejected promise with message when validation fails', () => {
      const wrapper = shallow(
        <Validator validators={failedValidator}>
          {({ validate }) => (
            <input
              type="text"
              onBlur={() => {
                validate().then(message => {
                  expect(message).toBe('message');
                });
              }}
            />
          )}
        </Validator>
      );

      wrapper.find('input').simulate('blur');
    });

    test('calls #onValidate when in ValidationProvider context', () => {
      const validator = { onValidate: jest.fn() };

      const wrapper = shallow(
        <Validator validators={trueValidator}>
          {({ validate }) => (
            <input
              type="text"
              onBlur={() => {
                validate().then(val => {
                  expect(val).toBeTruthy();
                  expect(validator.onValidate).toHaveBeenCalled();
                  expect(validator.onValidate.mock.calls[0][1]).toBeTruthy();
                });
              }}
            />
          )}
        </Validator>,
        { context: { validator } }
      );

      wrapper.find('input').simulate('blur');
    });

    test('sets state.isValid to false and message to failure message if validations fail', () => {
      const wrapper = shallow(
        <Validator validators={failedValidator}>
          {() => <div className="test-child" />}
        </Validator>
      );

      wrapper.instance().validate().then(() => {
        expect(wrapper.state().isValid).toBeFalsy();
        expect(wrapper.state().message).toBe('message');
      });
    });

    test('sets state.lastValidatedValue to the latest validated value', () => {
      const wrapper = shallow(
        <Validator validators={trueValidator}>
          {() => <div className="test-child" />}
        </Validator>
      );

      wrapper.instance().validate('test').then(() => {
        expect(wrapper.state().lastValidatedValue).toBe('test');
      });
    });

    test('uses state.lastValidatedValue if no value is provided in validate', () => {
      const mockValidator = jest.fn();
      mockValidator.mockReturnValueOnce(true).mockReturnValueOnce(true);

      const wrapper = shallow(
        <Validator validators={mockValidator}>
          {() => <div className="test-child" />}
        </Validator>
      );

      const testVal = 'test';

      wrapper.instance().validate(testVal).then(() => {
        wrapper.instance().validate().then(() => {
          expect(mockValidator).toBeCalledWith(testVal);
          expect(mockValidator).lastCalledWith(testVal);
        });
      });
    });
  });

  test('sets _mounted to true on mount', () => {
    const wrapper = mount(
      <Validator>
        {() => <div className="test-child" />}
      </Validator>
    );

    expect(wrapper.instance()._mounted).toBeTruthy();
  });

  test('sets _mounted to false on unmount', () => {
    const wrapper = shallow(
      <Validator>
        {() => <div className="test-child" />}
      </Validator>
    );

    const inst = wrapper.instance();
    wrapper.unmount();
    expect(inst._mounted).toBeFalsy();
  });

  test('#defaultState', () => {
    const wrapper = shallow(
      <Validator>
        {() => <div className="test-child" />}
      </Validator>
    );

    expect(wrapper.instance().defaultState()).toMatchSnapshot();
  });

  test('#clearValidation', () => {
    const wrapper = shallow(
      <Validator>
        {() => <div className="test-child" />}
      </Validator>
    );

    wrapper.setState({ isValid: false, message: 'blah' });
    wrapper.instance().clearValidation();
    expect(wrapper.state()).toEqual(wrapper.instance().defaultState());
  });

  test('#validateOnMount', () => {
    const mockValidatorOne = jest.fn();
    const mockValidatorTwo = jest.fn();

    mount(
      <Validator validators={mockValidatorOne} validateOnMount>
        {() => <div className="test-child" />}
      </Validator>
    );

    mount(
      <Validator validators={mockValidatorTwo}>
        {() => <div className="test-child" />}
      </Validator>
    );

    expect(mockValidatorOne.mock.calls.length).toBe(1);
    expect(mockValidatorTwo.mock.calls.length).toBe(0);
  });

  describe('In ValidationProvider context', () => {
    test('sets this.validator if validator is in context', () => {
      const validator = { registerValidator: () => {} };
      const wrapper = mount(
        <Validator>
          {() => <div className="test-child" />}
        </Validator>,
        { context: { validator } }
      );

      expect(wrapper.instance().validator).toEqual(validator);
    });

    test('calls registerValidator on mount and to set this.unsubscribe', () => {
      const validator = { registerValidator: jest.fn(), onValidate: () => {} };
      validator.registerValidator.mockReturnValueOnce('unsubscribe');

      const mockValidator = jest.fn();
      mockValidator.mockReturnValueOnce(true);

      const wrapper = mount(
        <Validator id="id" validators={mockValidator}>
          {() => <div className="test-child" />}
        </Validator>,
        { context: { validator } }
      );

      validator.registerValidator.mock.calls[0][1]();

      expect(validator.registerValidator).toHaveBeenCalled();
      expect(validator.registerValidator.mock.calls[0][0]).toBe('id');
      expect(mockValidator).toHaveBeenCalled();
      expect(wrapper.instance().unsubscribe).toBe('unsubscribe');
    });

    test('#onValide', () => {
      const validator = { onValidate: jest.fn() };

      const wrapper = shallow(
        <Validator id="id">
          {() => <div className="test-child" />}
        </Validator>,
        { context: { validator } }
      );

      wrapper.instance().onValidate();

      expect(validator.onValidate).toHaveBeenCalled();
      expect(validator.onValidate).toHaveBeenCalledWith('id', true);
    });
  });
});
