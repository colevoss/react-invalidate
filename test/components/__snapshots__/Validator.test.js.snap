// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Validator /> #defaultState 1`] = `
Object {
  "isValid": true,
  "lastValidatedValue": undefined,
  "message": null,
}
`;

exports[`<Validator /> passes the correct params to child function 1`] = `
Object {
  "clearValidation": [Function],
  "isValid": true,
  "message": null,
  "validate": [Function],
}
`;

exports[`<Validator /> renders it child function's component directly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="test-child"
/>,
  "nodes": Array [
    <div
      className="test-child"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Validator>
        [Function]
</Validator>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Validator {
        "_reactInternalInstance": [Circular],
        "clearValidation": [Function],
        "context": Object {
          "validator": undefined,
        },
        "id": undefined,
        "props": Object {
          "children": [Function],
        },
        "refs": Object {},
        "state": Object {
          "isValid": true,
          "lastValidatedValue": undefined,
          "message": null,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "validate": [Function],
        "validator": undefined,
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="test-child"
/>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="test-child"
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Validator>
    [Function]
</Validator>,
}
`;
